curp/src/client/state.rs

```diff
@@ -0,0 +1,322 @@
+use std::{
+    cmp::Ordering,
+    collections::{hash_map::Entry, HashMap, HashSet},
+    sync::Arc,
+};
+
+use event_listener::Event;
+use futures::{stream::FuturesUnordered, Future};
+use tokio::sync::RwLock;
+use tracing::{debug, info};
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        self,
+        connect::{BypassedConnect, ConnectApi},
+        CurpError, FetchClusterResponse, Protocol,
+    },
+};
+
+/// The client state
+#[derive(Debug)]
+pub(super) struct State {
+    /// Mutable state
+    mutable: RwLock<StateMut>,
```

Comment:
We could use `parking_lot::RwLock` for this.

Reply:
What are the benefits?

Reply:
It seems not easy to convert it into a synchronous lock. During `State::check_and_update`, there are asynchronous behaviors such as `ConnectApi::update_addrs` and `rpc::connects`. We cannot let the synchronous lock span across `.await`, as this may lead to recursive deadlock.